### PR TITLE
Implement auto push

### PR DESCRIPTION
**Pull request by FelonEkonom**

### CODE REVIEW

File: lib/membrane/core/element/toilet.ex [hunk 1]
```diff
@@ -75,57 +83,142 @@ defmodule Membrane.Core.Element.Toilet do
...
 
-    toilet_ref = DistributedCounter.new()
     capacity = capacity || default_capacity
-    {__MODULE__, toilet_ref, capacity, responsible_process, throttling_factor, 0}
+
+    receiver_effective_flow_control =
+      receiver_effective_flow_control
+      |> DistributedEffectiveFlowControl.new()
+
+    %__MODULE__{
```
Comment (varsill) [inline lib/membrane/core/element/toilet.ex:176]: Finally we have a struct for the toilet! 🥇 🥇 🥇

File: test/membrane/integration/effective_flow_control_resolution_test.exs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    Process.sleep(1000)
+
+    for idx <- 0..10, filter_type <- [:filter_a, :filter_b] do
+      child = {filter_type, idx}
+      child_pid = Testing.Pipeline.get_child_pid!(pipeline, child)
+      child_state = :sys.get_state(child_pid)
+
+      expected = if filter_type == :filter_a, do: :push, else: :pull
+      assert child_state.effective_flow_control == expected
+    end
```
Comment (varsill) [inline test/membrane/integration/effective_flow_control_resolution_test.exs:78]: you are repeating the code from lines 51-60. How about adding a private function to fetch the child's effective_flow_control?

**Review by varsill — CHANGES_REQUESTED** [review]
I have one general suggestion - perhaps we could try to somehow "merge" the process of resolving effective flow control with the process of resolving demand unit? Both take place in a similar place (PadController) and require similar data

File: lib/membrane/core/element/effective_flow_control_controller.ex [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+         %{direction: :input, flow_control: :auto} <- pad_data,
+         mode when mode in [:push, :pull] <- other_effective_flow_control do
+      resolve_effective_flow_control(state)
+    else
+      _other -> state
+    end
+  end
+
+  @spec resolve_effective_flow_control(State.t()) :: State.t()
+  def resolve_effective_flow_control(%State{effective_flow_control: :undefined} = state) do
```
[the remark this answers is not in the document]
Comment (FelonEkonom) [reply lib/membrane/core/element/effective_flow_control_controller.ex]: Exactly

**Comment by FelonEkonom** [comment]
Changes from this PR will be merged in https://github.com/membraneframework/membrane_core/pull/537

After discussion with @varsill we came to the conclusion, that it is better to not merge demand unit resolution with resolving effective flow control

**Pull request closed without merging**